### PR TITLE
[SPARK-35278][SQL] Invoke should find the method with correct number of parameters

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch fixes `Invoke` expression when the target object has more than one method with the given method name.

### Why are the changes needed?

`Invoke` will find out the method on the target object with given method name. If there are more than one method with the name, currently it is undeterministic which method will be used. We should add the condition of parameter number when finding the method.

### Does this PR introduce _any_ user-facing change?

Yes, fixed a bug when using `Invoke` on a object where more than one method with the given method name.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @maropu

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Nice catch!
Thank you, @viirya .

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (sunchao) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:329]: should we check parameter types too? also curious why we don't use `cls.getMethod` for the purpose.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:329]: +1 to use `getMethod`

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Actually `Invoke` uses `getMethod` for target object with data types other than `ObjectType`. I'm not sure why we need to do a separate `getMethod` during evaluation for the cases.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: And note `method` uses `encodedFunctionName` to find method, but in `eval` the original `functionName` is used.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: Switched to use `getMethod` and see what CI tells.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: Can we move the code to a method in the base class, and share it with `StaticInvoke`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: There is a bit difference between two `method` from `Invoke` and `StaticInvoke`.  For example, `StaticInvoke` doesn't use encoded function name. Is it safe to share `method` between them?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: We may backport this change, I'd like make sure we don't introduce regression unintentionally.

**Review by dongjoon-hyun — COMMENTED** [review]
BTW, @viirya . According to the `Affected Version` of SPARK-35278, are we going to cut another RC (Apache Spark 2.4.8 RC4) for this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: I see. Let's unify them later in the master branch only. Ideally, these 2 invoke functions should be consistent.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala]: It looks like a very ancient bug from Spark 1.x.

**Comment by viirya** [comment]
```
Caused by: sbt.ForkMain$ForkError: java.lang.RuntimeException: Couldn't find apply on interface scala.Function1
	at scala.sys.package$.error(package.scala:30)
	at org.apache.spark.sql.catalyst.expressions.objects.Invoke.liftedTree1$1(objects.scala:333)
	at org.apache.spark.sql.catalyst.expressions.objects.Invoke.method$lzycompute(objects.scala:329)
	at org.apache.spark.sql.catalyst.expressions.objects.Invoke.method(objects.scala:327)
	at org.apache.spark.sql.catalyst.expressions.objects.Invoke.doGenCode(objects.scala:359)
	at org.apache.spark.sql.catalyst.expressions.Expression.$anonfun$genCode$3(Expression.scala:147)
```

`getMethod` seems causing some test failures.

**Comment by viirya** [comment]
> BTW, @viirya . According to the `Affected Version` of [SPARK-35278](https://issues.apache.org/jira/browse/SPARK-35278), are we going to cut another RC (Apache Spark 2.4.8 RC4) for this?

Ideally, yes. Though seems no one complains this, and this is an ancient bug...

**Comment by viirya** [comment]
Let me know if you prefer to skip 2.4 backport.

**Comment by viirya** [comment]
Oh, the problem of using `getMethod` is for some cases, the parameter types is not known in advance. Like the failed tests, the argument types from `arguments` is `java.lang.Object`. So we cannot reliably using name + parameter types to get method.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: shall we use `filter` instead of `find`, and fail if more than one match is found?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: How about:

1. `getMethod` with param types
2. If not found, turn to find one method with parameter number
3. If more than one is found, or no one can be found, fails it

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: SGTM

**Review by maropu — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: +1 for that improved one.

**Comment by dongjoon-hyun** [comment]
> Let me know if you prefer to skip 2.4 backport.

For that, it's totally up to your decision, @viirya , because you are the release manager of Apache Spark 2.4.8. I support your decision.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:328]: +1. Now I'm wondering whether `StaticInvoke` has the same issue since it uses `getDeclaredMethod`.

**Comment by viirya** [comment]
This is weird:

```
sbt.ForkMain$ForkError: java.lang.RuntimeException: Error while encoding: java.lang.RuntimeException: Found 2 serialize on class org.apache.spark.ml.linalg.VectorUDT
newInstance(class org.apache.spark.ml.linalg.VectorUDT).serialize AS value#27
```

Found methods on `org.apache.spark.ml.linalg.VectorUDT` are:

```
method: serialize, WrappedArray(interface org.apache.spark.ml.linalg.Vector)                                                                                                                                                           
method: serialize, WrappedArray(class java.lang.Object) 
```

But `VectorUDT` only overrides `override def serialize(obj: Vector): InternalRow`. Why there is another `serialize(Object)`?

**Comment by viirya** [comment]
I think it is due to generic overloading.

**Comment by viirya** [comment]
Some changes were made after approvals. Please take another look before merging. cc @cloud-fan @dongjoon-hyun @maropu

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:342]: We cannot filter out synthetic ones in L336?

**Comment by maropu** [comment]
Looks fine otherwise.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala:342]: I don't know if it is possible that there is only one method and it is also a synthetic one. If we filter synthetic ones at L336, we may miss it?

**Comment by viirya** [comment]
Thanks @cloud-fan @dongjoon-hyun @maropu. Merging to master/3.1/3.0/2.4.

**Comment by dongjoon-hyun** [comment]
+1, late LGTM with the last commit.

**Comment by cloud-fan** [comment]
@viirya can we look into which callers do not provide the concrete parameter types when building `Invoke`? I think it's better to fix them, so that we can do a proper method resolution in `Invoke`, as a Java class may have several methods with the same name and number of arguments, but different argument types.

**Comment by viirya** [comment]
@cloud-fan Okay. Let me find them out and fix them.

**Comment by viirya** [comment]
@cloud-fan Just to confirm. So you mean we only allow `Invoke` to run the method with exact parameter type matches, right?

For example, calling a method `func(input: Object)` with argument `Tuple2` will be disallowed.

**Comment by viirya** [comment]
@cloud-fan Seems to me we still need to support it. One example is `UserDefinedType`.

For example, `VectorUDT` extends `UserDefinedType[Vector]`. `VectorUDT` is responsible for serialization of dense and sparse vectors. It overrides `def serialize(obj: Vector)` that takes both `SparseVector` and `DenseVector` as input parameters.

**Comment by cloud-fan** [comment]
> For example, calling a method func(input: Object) with argument Tuple2 will be disallowed.

This should be allowed, because we do provide the concrete parameter type (`Tuple2`). What should be forbidden is something like `func(s: String)` and the parameter type is `Object`.

I think I get your point, we should probably use a util function to look up the method, instead of writing the code by ourselves, to handle cases like search the inheritance hierarchy, argument type casting, etc.

How about https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/reflect/MethodUtils.html#getMatchingMethod-java.lang.Class-java.lang.String-java.lang.Class...-

**Comment by viirya** [comment]
> > For example, calling a method func(input: Object) with argument Tuple2 will be disallowed.
> 
> This should be allowed, because we do provide the concrete parameter type (`Tuple2`). What should be forbidden is something like `func(s: String)` and the parameter type is `Object`.
> 
> I think I get your point, we should probably use a util function to look up the method, instead of writing the code by ourselves, to handle cases like search the inheritance hierarchy, argument type casting, etc.
> 
> How about https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/reflect/MethodUtils.html#getMatchingMethod-java.lang.Class-java.lang.String-java.lang.Class...-

It looks promising. Not aware there is an util function we can use. So my question would be should we use it in backports? Or only replacing with our code with the util in master only?

**Comment by cloud-fan** [comment]
I think this PR itself is pretty conservative and is good to backport. The new code is clearly better than the old code without regression. I'd suggest to only change to the util method in the master branch, for both invoke and static invoke.

**Comment by cloud-fan** [comment]
BTW, AFAIK invoke and static invoke are only used by internal places such as the encoder and UDT. It's not really a bug for <= 3.1, but a bug for 3.2 as we use it to execute the new UDF.

**Comment by viirya** [comment]
> I think this PR itself is pretty conservative and is good to backport. The new code is clearly better than the old code without regression. I'd suggest to only change to the util method in the master branch, for both invoke and static invoke.

I see. Then I don't touch invoke in old branches again and keep this backport as it is. Let's use the util method in master only.

**Comment by viirya** [comment]
> BTW, AFAIK invoke and static invoke are only used by internal places such as the encoder and UDT. It's not really a bug for <= 3.1, but a bug for 3.2 as we use it to execute the new UDF.

Rarely but maybe possibly developers will use invoke/static invoke for custom encoder? As this is already backport to 2.4/3.0/3.1, I think it should be fine to keep as it is? For static invoke #32413, it is also conservative, so I guess it may be also good for backport?

**Comment by cloud-fan** [comment]
> For static invoke #32413, it is also conservative, so I guess it may be also good for backport?

Yea I think so

**Pull request closed without merging**
